ArduPlane/navigation.cpp

```diff
@@ -152,19 +156,47 @@ void Plane::calc_airspeed_errors()
 
 void Plane::calc_gndspeed_undershoot()
 {
- 	// Use the component of ground speed in the forward direction
-	// This prevents flyaway if wind takes plane backwards
-    if (gps.status() >= AP_GPS::GPS_OK_FIX_2D) {
-	    Vector2f gndVel = ahrs.groundspeed_vector();
-		const Matrix3f &rotMat = ahrs.get_rotation_body_to_ned();
-		Vector2f yawVect = Vector2f(rotMat.a.x,rotMat.b.x);
-        if (!yawVect.is_zero()) {
-            yawVect.normalize();
-            float gndSpdFwd = yawVect * gndVel;
-            groundspeed_undershoot = (aparm.min_gndspeed_cm > 0) ? (aparm.min_gndspeed_cm - gndSpdFwd*100) : 0;
-        }
+
+    // Use the component of ground speed in the forward direction
+    // This prevents flyaway if wind takes plane backwards
+    const Matrix3f &rotMat = ahrs.get_rotation_body_to_ned();
+    Vector2f yawVect = Vector2f(rotMat.a.x,rotMat.b.x);
+
+    // Use integrator based gndspd_undershoot controller
+    if (gps.status() >= AP_GPS::GPS_OK_FIX_2D &&
+        is_flying() && !yawVect.is_zero() &&
+        (aparm.min_gndspeed_cm > 0)) {
+
+        float nav_bearing = radians(0.01f*nav_controller->target_bearing_cd());
+        Vector2f navVect = Vector2f(cosf(nav_bearing), sinf(nav_bearing));
+        navVect.normalize();
+
+        Vector2f gndSpdVect = ahrs.groundspeed_vector();
+        yawVect.normalize();
+
+        // Using the component of ground speed in the forward direction only
+        // prevents flyaway if wind takes plane backwards, but may
+        // result in underestimating the actual ground speed along track.
+        // Hence, use the max of both estimates.
+
+        float gndspd_forward     = yawVect * gndSpdVect;
+        float gndspd_along_track = navVect * gndSpdVect;
+        float gndSpdFwd = MAX(gndspd_forward, gndspd_along_track);
+
+        float gndspd_undershoot_error = aparm.min_gndspeed_cm - gndSpdFwd*100;
+        // compensate for the lag in airspeed controller (coverting to cm/sec)
+        gndspd_undershoot_error -= airspeed_error*100;
```

Comment:
Hi, I've made it like that since otherwise I am getting oscillations on the demand whenever the airspeed controller is off its target. This is basically warranting that the actual underspeed is not related to lagging behind the target airspeed.